src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -232,28 +244,34 @@ public async Task<bool> WaitFor100ContinueAsync(CancellationToken cancellationTo
                 return sendRequestContent;
             }
 
-            private void CheckForCompletion()
+            private void Reset()
             {
                 Debug.Assert(Monitor.IsEntered(SyncObject));
-                Debug.Assert(_requestCompletionState != StreamCompletionState.InProgress || _responseCompletionState != StreamCompletionState.InProgress,
-                    $"CheckForCompletion called but neither request nor response is completed");
-
-                if (_requestCompletionState == StreamCompletionState.InProgress || _responseCompletionState == StreamCompletionState.InProgress)
+                Debug.Assert(_requestCompletionState != StreamCompletionState.InProgress);
+                Debug.Assert(_responseCompletionState != StreamCompletionState.InProgress);
+                Debug.Assert(_requestCompletionState == StreamCompletionState.Failed || _responseCompletionState == StreamCompletionState.Failed,
+                    "Reset called but neither request nor response is failed");
+
+                if (NetEventSource.IsEnabled) Trace($"Stream reset. This is a test. Request={_requestCompletionState}, Response={_responseCompletionState}.");
+                // Don't send a RST_STREAM if we've already received one from the server.
+                if (_resetException == null)
                 {
-                    return;
+                    _connection.LogExceptions(_connection.SendRstStreamAsync(_streamId, Http2ProtocolErrorCode.Cancel));
```

Comment:
Is Http2ProtocolErrorCode.Cancel always the right error code to send?  It looks like we might send this if we experienced some kind of error; should that be represented to the server as Cancel?  I'm fine if the answer is "yes", just want to make sure.

Reply:
https://tools.ietf.org/html/rfc7540#section-7

The only other error that I can see might apply here is INTERNAL_ERROR.

But honestly, I don't think it matters much.